core/src/block.rs

```diff
@@ -187,6 +188,16 @@ impl<'x> OpenBlock<'x> {
         Ok(())
     }
 
+    pub fn prepare_block_from_transactions<'a>(
+        &mut self,
+        block_executor: &dyn BlockExecutor,
+        transactions: impl Iterator<Item = &'a TransactionWithMetadata>,
+    ) {
+        let transactions = transactions.collect::<Vec<_>>();
```

Comment:
If you change `prepare_block` to receive the iterator also, you don't need to create the vector in here.

Reply:
That's true but we still need something complex here, because we have to copy or reference the given iterator somehow because of lifetime parameters of `prepare_block` require it.

Reply:
You don't need to pass the `Iterator` as `&mut`, `Box<dyn Iterator<Item = &'a TransactionWithMetadata> + 'a>` is enough.
If you think it's complex yet, you can make the `'a` and the Iterator parameters of `BlockExecutor`, like 
```rust
pub trait BlockExecutor<'a, Transactions>: Send + Sync  where Transactions: Iterator<Item = &'a TransactionWithMetadata>
```
But it will make the other code that uses the trait more complex.

Reply:
Oh, I didn't think about using a lifetime keyword in the `Box` type. Thanks, I'm still getting used to it.
I actually thought about the latter option, but then I couldn't instantiate `dyn BlockExecutor` object.

Reply:
I just realized that I could just use `&mut dyn Iterator` as well.
But I think `Box<dyn Iterator>` is better than `&mut dyn Iterator`, since it states about the reusability more clearly.
I'll use `Box<dyn Iterator>` for both `prepare_block` and `filter_transactions`.